pkg/engine/lifecycletest/update_plan_test.go

```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
+		}
+
+		program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+				Inputs: resource.PropertyMap{
+					"foo": resource.NewStringProperty("bar"),
+				},
+			})
+			assert.NoError(t, err)
+
+			_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+				Inputs: resource.PropertyMap{
+					"foo": resource.NewStringProperty("bar"),
+				},
+			})
+			assert.NoError(t, err)
+
+			return nil
+		})
+
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		return host
+	}()
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Define a function that creates new resource plans as resource plans get mutated during use.
+	getPlan := func() *deploy.Plan {
+		plan, res := TestOp(Update).Plan(project, p.GetTarget(t, nil), UpdateOptions{
+			Host:         host,
+			GeneratePlan: true,
+			Experimental: true,
+			UpdateTargets: deploy.NewUrnTargets([]string{
+				"urn:pulumi:test::test::pkgA:m:typA::resB",
+				"urn:pulumi:test::test::pulumi:providers:pkgA::default",
+			}),
+		}, p.BackendClient, nil)
+		assert.Nil(t, res)
+		assert.NotNil(t, plan)
+		return plan
+	}
+
+	// Check that running an update with everything targeted fails due to our plan being constrained
+	// to the resource and default provider.
+	{
+		_, res := TestOp(Update).Run(project, p.GetTarget(t, nil), UpdateOptions{
+			Plan:         getPlan(),
+			Host:         host,
+			Experimental: true,
+		}, false, p.BackendClient, nil)
+		assert.NotNil(t, res)
+	}
+
+	{
+		_, res := TestOp(Update).Run(project, p.GetTarget(t, nil), UpdateOptions{
+			Plan:         getPlan(),
+			Host:         host,
+			Experimental: true,
+			UpdateTargets: deploy.NewUrnTargets([]string{
+				"urn:pulumi:test::test::pkgA:m:typA::resB",
+				"urn:pulumi:test::test::pulumi:providers:pkgA::default",
+			}),
+		}, false, p.BackendClient, nil)
+		assert.Nil(t, res)
+	}
+}
+
+// TestTargetedDoesNotCreateProvider checks that a plan that targets a resource does not create the default
+// provider if not targeted.
+func TestTargetedDoesNotCreateProvider(t *testing.T) {
+	t.Parallel()
+
+	host := func() plugin.Host {
+		loaders := []*deploytest.ProviderLoader{
+			deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+				return &deploytest.Provider{}, nil
+			}),
+		}
+
+		program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{})
+			assert.NoError(t, err)
+
+			return nil
+		})
+
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		return host
+	}()
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Check that update fails because the default provider is not created (as it is not targeted).
```

Comment:
There's part of me a little worried about this. We don't generally make users think about default providers too much and this is going to require them to learn default provider URNs just to create new targetted resources.

I think the change "explicit provider resources shouldn't be created unless targeted" is fine, as even if people are currently relying on this behavior I'm much happier pointing out to them that this is like having a normal resource dependency and expecting it to work if you only targeted the dependent resource.